examples/supplementary/arg_iter/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+        let fn_sig = cx.tcx.fn_sig(id).skip_binder(); // Use skip_binder to handle late-bound lifetimes
+
+        for (i, input_ty) in fn_sig.inputs().iter().enumerate() {
+            // Check if the input type is a generic parameter (TyKind::Param)
+            if let ty::TyKind::Param(param_ty) = input_ty.skip_binder().kind() {
+                // Map the ParamTy index back to the GenericParamDef
+                // Use param_at since opt_param_at is deprecated and the index should be valid
+                let param_def = generics.param_at(param_ty.index as usize, cx.tcx);
+
+                // Check if this parameter's name is in our list of iterator-only params
+                if iterator_only_params.contains(&param_def.name) {
+                    // Get the span from the corresponding HIR FnDecl input type for accurate
+                    // reporting
+                    let input_hir_ty = &decl.inputs[i];
+                    span_lint_and_help(
+                        cx,
+                        ARG_ITER,
+                        input_hir_ty.span, /* Lint the type in the function signature (using
+                                            * HIR span) */
+                        "parameter type has `Iterator` bound",
+                        None,
+                        format!(
+                            "consider using `IntoIterator` instead of `Iterator` for parameter `{}`",
+                            param_def.name
+                        ),
+                    );
+                }
+            }
+        }
+    }
+}
+
+/// Checks if a given type parameter `param_def_id` is bound _only_ by `iterator_def_id`
+/// within the given `predicates`. Implicit `Sized` bounds are ignored.
+fn is_param_bound_only_by_iterator<'tcx>(
+    cx: &LateContext<'tcx>,
+    predicates: ty::GenericPredicates<'tcx>,
+    param_def_id: DefId, // The DefId of the type parameter we are checking (e.g., `I`)
+    iterator_def_id: DefId,
+) -> bool {
+    let Some(sized_def_id) = cx.tcx.lang_items().sized_trait() else {
+        return false; // Cannot check without Sized trait, assume false
+    };
+
+    // Get the owner ID (e.g., the function DefId) to correctly resolve ParamTy indices
+    let owner_id = cx.tcx.parent(param_def_id);
+
+    let mut found_iterator_bound = false;
+    let mut found_other_bound = false;
+
+    for &(predicate, _pred_span) in predicates.predicates.iter() {
+        // Prefix unused span
+        // We only care about trait bounds applied to our specific parameter
+        if let Some(trait_pred) = predicate.as_trait_clause() {
+            // Use skip_binder() for the trait predicate as well
+            let bound_trait_pred = trait_pred.skip_binder();
+            let self_ty = bound_trait_pred.self_ty();
+
+            // Check if the bound applies to our parameter `param_def_id`
+            if let ty::TyKind::Param(p_ty) = self_ty.kind() {
```

Comment:
Here, I think you should be able to just compare the passed-in `Ty` to `self_ty`.